### PR TITLE
Generate Flowtype definitions with a flag.

### LINKED ISSUE #11: Support Flow definitions generation ?
Hi,

I would like to contribute to this project with a pull request bringing Flow (https://flow.org/) definitions generation. At fluo.com we use protocol buffers too and we use Flow so it would be definitely be helpful !

I was thinking of adding a "_.bin/protoc-gen-flow_" plugin that would output flow definitions just like typescript with proper flow options.

How would you welcome such a PR ?

Regards,

Florent

### PR DESCRIPTION
**Pull request by chrisgervang**
Fixes #11 

Hi there, I've really enjoyed using this tool for typescript. I've recently been working on a project that uses flowtype, which also uses protobuf. So, I spent some time this week adding flowtype generation behind a `flow=true` flag. 

**Usage**
`--ts_out=flow=true:some/output/dir`
It still defaults to generating `ts`, and it should never generate both `flow` and `ts` definitions.

**Testing**
For testing, I've added a `flow` example and generated the output using the provided proto files. To test flow, just `npm install` and run `npm run flow` - we should see `No errors!`. I also have a large collection of proto files in our project, and it worked well for all of them. 

**Approach**
I noticed that while the structure of the code generation code remains similar, I needed to make a lot of changes to work around the lack of a `namespace` in flow. So, I created a `flow` source folder with modified versions of the `ts` folder sources. 

Additionally, flow doesn't support a pure definition file which means the types need to be manually applied to the `js` import, like so:
```
import {ExampleMessage} from 'example_pb.js';
import {ExampleMessage$AsClass} from 'example_pb.flow.js';

const rawData: any;
const example: ExampleMessage$AsClass = ExampleMessage.deserializeBinary(
    new Uint8Array(rawData)
);

// now example is typed and contains data.
```
This "gluing" of type and implementation could eventually be done automatically for each message type in the `flow.js` files, but is unnecessary to get all of the benefit of this tool. 

**Maintenance**
I'm happy to maintain flow related bugs that come up. I'd ideally like to keep these projects together, but if that isn't something the maintainers are interested in then I hope it's not a problem to fork as a new tool.

**Whats Next**
This doesn't yet support generating `service` code, though that could be added if needed. Also, `google-protobuf` doesn't have a `flowtype` definition yet, so I may add that to `flow-typed` and then later add that to the `flow` example project here. As a result, the type coverage isn't as high as it could be, but it's pretty good and still very helpful.

### CODE REVIEW

**Comment by trevorwhealy** [comment]
Nice work, cannot wait to use this.

**Comment by jonny-improbable** [comment]
Awesome work @chrisgervang; thank you so much for your contribution!

This change looks good, however I am concerned that we could easily regress flow-typings in the future as there doesn't appear to be any kind of automated test (although maybe I missed it) - with the TSD definitions we compile them as part of the integration test suite, would it be possible to do something similar for the generate flow types?

**Comment by chrisgervang** [comment]
For sure. Just wanted to see how this would be received first. I could implement them in a follow up PR, if that's alright with you. @trevorwhealy and @georgios-uber, would you be interested in helping me reimplement the TS tests for flow? 
https://github.com/improbable-eng/ts-protoc-gen/tree/master/test

**Comment by jonnyreeves** [comment]
Thanks @chrisgervang, more than happy to land this, although it will require some form of testing.

**Comment by chrisgervang** [comment]
Ok, I’ll work on some tests. I’ll likely create a flow test folder and reimplement many of the same tests that you’ve got for TSD files. Thank you for providing so many tests, and making the tool easy to extend!

**Comment by jonnyreeves** [comment]
Thanks, if your time is limited then feel free to do for a smaller slice of coverage and we can look to backfill. Happy to help if you have any questions

**Comment by chrisgervang** [comment]
@jonnyreeves I'm not sure what happened on travis. I made a mistake on a package-lock checkin, and now travis appears to be stuck using a version of package-lock that isn't in my PR anymore.

**Comment by chrisgervang** [comment]
Ah, disabling the travis cache feature fixes the issue. I'm not sure how to flush the cache. See: https://travis-ci.org/improbable-eng/ts-protoc-gen/builds/457321929?utm_source=github_status&utm_medium=notification

I'm going to enable it again.

**Comment by chrisgervang** [comment]
Looks like the cache is stale and needs to be deleted. Seems like this command my help (I don't imagine I have access to run it): https://github.com/travis-ci/travis.rb#cache

**Comment by jonnyreeves** [comment]
Sorry for not looking into this yet, I've been really snowed under with work. Hope to get onto this soon

**Comment by jonnyreeves** [comment]
@chrisgervang going to take a look now, but please can you merge in the latest from master and resolve the conflicts in the package.json files? Thanks!

File: .travis.yml [hunk 1]
```diff
@@ -27,9 +27,7 @@ deploy:
   tag: "$NPM_TAG"
   on:
     branch: master
-cache:
-  directories:
-  - node_modules
+cache: false
```
Comment (jonnyreeves) [inline .travis.yml]: Should be able to revert this now.

File: babel.config.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// used for flow type stripping and
```
Comment (jonnyreeves) [inline babel.config.js]: and....? :)

File: babel.config.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+module.exports = function(api) {
+  api.cache(false);
```
Comment (jonnyreeves) [inline babel.config.js]: why are we disabling the cache here? Please could you document if this is required.

File: flow-typed/npm/mocha_v5.x.x.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
+// flow-typed signature: ae94222964881f91f3333824878bbb2f
```
Comment (jonnyreeves) [inline flow-typed/npm/mocha_v5.x.x.js]: Should this have been committed to the repo?

File: package.json [hunk 5]
```diff
@@ -29,16 +30,22 @@
   "author": "Improbable",
   "license": "Apache-2.0",
   "dependencies": {
+    "@babel/polyfill": "^7.0.0",
```
Comment (jonnyreeves) [inline package.json]: Shouldn't this be a devDependency?

File: flow-typed/npm/mocha_v5.x.x.js (hunk 4 above)
Comment (chrisgervang) [reply flow-typed/npm/mocha_v5.x.x.js]: yes, unfortunately there isn't a tool like `@types` for flow.

File: package.json (hunk 5 above)
Comment (chrisgervang) [reply package.json]: I think I can remove this actually, but no it's the package that has polyfills used at runtime.

**Comment by chrisgervang** [comment]
Thanks for getting back to me @jonnyreeves. I'll get started on turning my fork into a proper `flow-protoc-gen` repo. Do you have any tips you can share with me about setting up the CI and repo maintenance tools you guys use on this repo?

**Comment by jonnyreeves** [comment]
Sure thing @chrisgervang; we use [TravisCI](https://travis-ci.com/getting_started) - they provide a free tier for OpenSource projects. You should be able to lift most of this project's [`.travis.yml`](https://github.com/improbable-eng/ts-protoc-gen/blob/master/.travis.yml) file as-is, although you may wish to remove the `before_deploy` and `deploy` steps (which publish to NPM on each commit) when you are first getting started.

Happy to answer any other questions you may have, but probably best if we continue the discussion over on your fork's issue tracker. Feel free to open a new issue and ping me if you have issues :)

hth

**Pull request closed without merging**